File: problems/106/main.cpp
#include <iostream>
#include <cstdio>
#include <algorithm>
#include <cstring>
#include <string>
#include <cctype>
#include <stack>
#include <queue>
#include <list>
#include <vector>
#include <map>
#include <sstream>
#include <cmath>
#include <bitset>
#include <utility>
#include <set>
#include <numeric>
#include <ctime>

using namespace std;

struct ListNode {
	int val;
	ListNode* next;
	ListNode(int x) : val(x), next(NULL) {}
};

class Solution {
	public:
		ListNode* mergeKLists(vector<ListNode*>& lists);
};

ListNode* mergeTwoLists(ListNode* a, ListNode* b) {
	if (!a) return b;
	if (!b) return a;
	if (a->val < b->val) {
		a->next = mergeTwoLists(a->next, b);
		return a;
	} else {
		b->next = mergeTwoLists(a, b->next);
		return b;
	}
}

ListNode* Solution::mergeKLists(vector<ListNode*>& lists) {
    if (lists.empty()) {
        return NULL;
    }
    queue<ListNode*> Q;
    for (int i = 0; i < lists.size(); i++) Q.push(lists[i]);
    ListNode* last = NULL;
    while(!Q.empty()) {
        ListNode* first = Q.front(); Q.pop();
        if (Q.empty()) return first;
        ListNode* second = Q.front(); Q.pop();
        ListNode* mergedList = mergeTwoLists(first, second);
        Q.push(mergedList);
        last = mergedList;
    }
    return last;
}

void addNodeToEnd(ListNode*& head, int val) {
	ListNode* tmp = new ListNode(val);
	if (head == NULL) {
		head = tmp;
		head->next = NULL;
	} else {
		ListNode* tail = head;
		while (tail->next != NULL) {
			tail = tail->next;
		}
		tail->next = tmp;
	}
}

void printLinkedList(ListNode* head) {
	if (!head) {
		cout << endl;
		return;
	}
	ListNode* tmp = head;
	while(tmp->next != NULL) {
		cout << tmp->val << " -> ";
		tmp = tmp->next;
	}
	cout << tmp->val << endl;
}

int main() {
	int testcases;
	cin >> testcases;

	for (int t = 0; t < testcases; t++) {
		int n; cin >> n;
		vector<ListNode*> lists;

		for (int i = 0; i < n; i++) {
			int m; cin >> m;
			ListNode* head = NULL;
			for (int j = 0; j < m; j++) {
				int x; cin>> x;
				addNodeToEnd(head, x);
			}
			lists.push_back(head);
		}

		Solution s;
		ListNode* mergedHead = s.mergeKLists(lists);
		printLinkedList(mergedHead);
	}
	return 0;
}

File: problems/106/Main.java
import java.io.*;
import java.util.*;

class ListNode {
    int val;
    ListNode next;
    ListNode(int x) {
        val = x;
        next = null;
    }
}
// ListNode type
// class ListNode {
//     int val;
//     ListNode next;
//     ListNode(int x) { val = x; next = null; }
// }

class Solution {
    public static ListNode mergeKLists(ListNode[] lists) {
        if(lists==null||lists.length==0)
            return null;

        PriorityQueue<ListNode> queue = new PriorityQueue<ListNode>(new Comparator<ListNode>(){
            public int compare(ListNode l1, ListNode l2){
                return l1.val - l2.val;
            }
        });

        ListNode head = new ListNode(0);
        ListNode p = head;

        for(ListNode list: lists){
            if(list!=null)
                queue.offer(list);
        }

        while(!queue.isEmpty()){
            ListNode n = queue.poll();
            p.next = n;
            p=p.next;

            if(n.next!=null)
                queue.offer(n.next);
        }

        return head.next;

    }
}


class LinkedList {
    ListNode head;
    ListNode tail;

    LinkedList() {
        this.head = null;
        this.tail = null;
    }

    public void addNodeToEnd(int val) {
        ListNode tmp = new ListNode(val);
        tmp.next = null;
        if (head == null) {
            head = tmp;
            head.next = null;
            tail = head;
        } else {
            tail.next = tmp;
            tail = tail.next;
        }
    }

    public void printLinkedList() {
        if (head == null) {
            System.out.println();
            return;
        }

        ListNode tmp = head;
        while(tmp.next != null) {
                System.out.print(tmp.val);
                System.out.print(" -> ");
            tmp = tmp.next;
        }
        System.out.println(tmp.val);
    }


}

public class Main {

    public static void main(String args[]) throws Exception {
        BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));
        int cases = Integer.parseInt(reader.readLine());

        for (int i = 1; i <= cases; i++) {
            String inputLine = reader.readLine();
            String inputItems[] = inputLine.split(" ");

            int x = 0;
            int numLists = Integer.parseInt(inputItems[0]);
            x++;
            ListNode[] lists = new ListNode[numLists];

            for (int j = 0; j < numLists; j++) {
                int numNodes = Integer.parseInt(inputItems[x]);
                x++;
                LinkedList L = new LinkedList();
                while (numNodes > 0) {
                    L.addNodeToEnd(Integer.parseInt(inputItems[x]));
                    x++;
                    numNodes--;
                }
                lists[j] = L.head;
            }

            Solution s = new Solution();
            LinkedList resultList = new LinkedList();
            resultList.head = s.mergeKLists(lists);
            resultList.printLinkedList();
        }
    }
}
